packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts

```diff
@@ -0,0 +1,366 @@
+  public static grpc(props: GrpcNodeListenerOptions = {}): VirtualNodeListener {
+    return new GrpcNodeListener(props);
+  }
+
+  /**
+   * Returns an TCP Listener for a VirtualNode
+   */
+  public static tcp(props: TcpNodeListenerOptions = {}): VirtualNodeListener {
+    return new TcpNodeListener(props);
+  }
+
+  /**
+   * Protocol the listener implements
+   */
+  protected abstract protocol: Protocol;
+
+  /**
+   * Port to listen for connections on
+   */
+  protected abstract port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   */
+  protected abstract healthCheck?: HealthCheck;
+
+  /**
+   * Represents Listener Timeout
+   */
+  protected abstract timeout?: HttpTimeout;
+
+  /**
+   * Binds the current object when adding Listener to a VirtualNode
+   */
+  public bind(_scope: cdk.Construct): VirtualNodeListenerConfig {
+    return {
+      listener: {
+        portMapping: {
+          port: this.port,
+          protocol: this.protocol,
+        },
+        healthCheck: this.healthCheck ? this.renderHealthCheck(this.healthCheck) : undefined,
+        timeout: this.timeout ? this.renderTimeout(this.timeout) : undefined,
+      },
+    };
+  }
+
+  /**
+   * Returns an HealthCheck for a VirtualNode
+   */
+  protected renderHealthCheck(hc: HealthCheck): CfnVirtualNode.HealthCheckProperty | undefined {
+    if (hc === undefined) { return undefined; }
+
+    if (hc.protocol === Protocol.TCP && hc.path) {
+      throw new Error('The path property cannot be set with Protocol.TCP');
+    }
+
+    if (hc.protocol === Protocol.GRPC && hc.path) {
+      throw new Error('The path property cannot be set with Protocol.GRPC');
+    }
+
+    const healthCheck: CfnVirtualNode.HealthCheckProperty = {
+      healthyThreshold: hc.healthyThreshold || 2,
+      intervalMillis: (hc.interval || cdk.Duration.seconds(5)).toMilliseconds(), // min
+      path: hc.path || (hc.protocol === Protocol.HTTP ? '/' : undefined),
+      port: hc.port || this.port,
+      protocol: hc.protocol || this.protocol,
+      timeoutMillis: (hc.timeout || cdk.Duration.seconds(2)).toMilliseconds(),
+      unhealthyThreshold: hc.unhealthyThreshold || 2,
+    };
+
+    validateHealthChecks(healthCheck);
+
+    return healthCheck;
+  }
+
+  /**
+   * Returns the ListenerTimeoutProperty based on protocol
+   */
+  private renderTimeout(timeout: HttpTimeout): CfnVirtualNode.ListenerTimeoutProperty {
+    return ({
+      [this.protocol]: {
+        idle: timeout?.idle !== undefined ? {
+          unit: 'ms',
+          value: timeout?.idle.toMilliseconds(),
+        } : undefined,
+        perRequest: timeout?.perRequest !== undefined ? {
+          unit: 'ms',
+          value: timeout?.perRequest.toMilliseconds(),
+        } : undefined,
+      },
+    });
+  }
+}
+
+/**
+ * Represents the properties required to define a HTTP Listener for a VirtualNode.
+ */
+class HttpNodeListener extends VirtualNodeListener {
```

Comment:
There's no need for these subclasses. They don't have any logic inside of them, except copy-pasted defaulting to port 8080. Here's how to get rid of all this.

```ts
/**
 *  Defines listener for a VirtualNode
 */
export abstract class VirtualNodeListener {
  /**
   * Returns an HTTP Listener for a VirtualNode
   */
  public static http(props: HttpNodeListenerOptions = {}): VirtualNodeListener {
    return new VirtualNodeListenerImpl(Protocol.HTTP, props.healthCheck, props.timeout, props.port);
  }

  /**
   * Returns an HTTP2 Listener for a VirtualNode
   */
  public static http2(props: HttpNodeListenerOptions = {}): VirtualNodeListener {
    return new VirtualNodeListenerImpl(Protocol.HTTP2, props.healthCheck, props.timeout, props.port);
  }

  /**
   * Returns an GRPC Listener for a VirtualNode
   */
  public static grpc(props: GrpcNodeListenerOptions = {}): VirtualNodeListener {
    return new VirtualNodeListenerImpl(Protocol.GRPC, props.healthCheck, props.timeout, props.port);
  }

  /**
   * Binds the current object when adding Listener to a VirtualNode
   */
  public abstract bind(scope: cdk.Construct): VirtualNodeListenerConfig;
}

// module-private
class VirtualNodeListenerImpl extends VirtualNodeListener {
  constructor(private readonly protocol: Protocol, 
    private readonly healthCheck: HealthCheck | undefined,
    private readonly timeout: HttpTimeout | undefined,
    private readonly port: number = 8080) {
  }

  public bind(_scope: cdk.Construct): VirtualNodeListenerConfig {
    return {
      listener: {
        portMapping: {
          port: this.port,
          protocol: this.protocol,
        },
        healthCheck: this.healthCheck ? this.renderHealthCheck(this.healthCheck) : undefined,
        timeout: this.timeout ? this.renderTimeout(this.timeout) : undefined,
      },
    };
  }

  private renderHealthCheck(hc: HealthCheck): CfnVirtualNode.HealthCheckProperty | undefined {
    if (hc === undefined) { return undefined; }

    if (hc.protocol === Protocol.TCP && hc.path) {
      throw new Error('The path property cannot be set with Protocol.TCP');
    }

    if (hc.protocol === Protocol.GRPC && hc.path) {
      throw new Error('The path property cannot be set with Protocol.GRPC');
    }

    const healthCheck: CfnVirtualNode.HealthCheckProperty = {
      healthyThreshold: hc.healthyThreshold || 2,
      intervalMillis: (hc.interval || cdk.Duration.seconds(5)).toMilliseconds(), // min
      path: hc.path || (hc.protocol === Protocol.HTTP ? '/' : undefined),
      port: hc.port || this.port,
      protocol: hc.protocol || this.protocol,
      timeoutMillis: (hc.timeout || cdk.Duration.seconds(2)).toMilliseconds(),
      unhealthyThreshold: hc.unhealthyThreshold || 2,
    };

    validateHealthChecks(healthCheck);

    return healthCheck;
  }

  private renderTimeout(timeout: HttpTimeout): CfnVirtualNode.ListenerTimeoutProperty {
    return ({
      [this.protocol]: {
        idle: timeout?.idle !== undefined ? {
          unit: 'ms',
          value: timeout?.idle.toMilliseconds(),
        } : undefined,
        perRequest: timeout?.perRequest !== undefined ? {
          unit: 'ms',
          value: timeout?.perRequest.toMilliseconds(),
        } : undefined,
      },
    });
  }
}